### ISSUE TITLE
intent for compiler-generated array arguments in wrapper functions

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#15917]
PR #15903 adjusts the compiler so that an `export proc` argument with `in` intent will be passed by value from C code. However it applies a workaround to `export proc` arguments generated in wrapping `export` functions with array arguments for Python interop.

I think that the Python interop should build from the same starting points as the C interop - and in particular a reason for that is that we can't know at compile time whether an `export proc` will be used from C or Python. As a result, I think these wrapper functions should apply also to C and should use the natural by-value C passing style.

**Comment by bradcray** [comment]
> we can't know at compile time whether an export proc will be used from C or Python.

I'm not sure that's right...  At present, I think the model is that `--library-python` says you're creating a Chapel library that will be called from Python (and not C), `--library-fortran` says you're creating one that will be called from Fortran, and (asymmetrically), `--library` creates one for C.

**Comment by mppf** [comment]
> > we can't know at compile time whether an export proc will be used from C or Python.
> 
> I'm not sure that's right...  At present, I think the model is that `--library-python` says you're creating a Chapel library that will be called from Python (and not C), `--library-fortran` says you're creating one that will be called from Fortran, and (asymmetrically), `--library` creates one for C.

Yes, I know. But consider a Chapel file like this that is compiled with `--library-python`:

``` chapel
extern {
  // using extern block for short example; same thing possible without one
  void mychapelfunc(void);
  void mycfunc(void) {
    mychapelfunc();
  }
}


mycfunc(); // calls mychapelfunc from C
// mychapelfunc might also be called from Python when this is compiled with --library-python

export proc mychapelfunc() {
}
```

Here `mychapelfunc` is called from C (and also possibly from Python).

**Comment by mppf** [comment]
Issue #16134 proposes a solution to Python-specific or multilocale-specific behavior for `export proc`s.
